### PR TITLE
Fix bug that a callback not reached

### PR DESCRIPTION
**Pull request by iuhilnehc-ynos**
a subscriber on a new executor with a callback group triggered to receive a message

related to https://github.com/ros2/rclcpp/issues/1611

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

### CODE REVIEW

**Comment by iuhilnehc-ynos** [comment]
this test case expects test result as

```
The following tests FAILED:
	 26 - test_add_callback_groups_to_executor__rmw_fastrtps_cpp (Failed)
	 27 - test_add_callback_groups_to_executor__rmw_fastrtps_dynamic_cpp (Failed)
```

after using https://github.com/ros2/rmw_fastrtps/pull/527, there is no failure.

**Review by wjwwood — APPROVED** [review]

**Review by fujitatomoya — APPROVED** [review]

**Comment by asorbini** [comment]
Before merging this test, can we discuss whether attaching a node to two different executors should be allowed by the API? This isn't currently a requirement AFAIK, so making this change is not trivial at all.

File: rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp
Comment (wjwwood) [inline rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp:324]: When `subscription` and `publisher` go out of scope they will be removed from the graph, and the 100ms sleep is unlikely to be enough all the time, i.e. this will be flaky.

Is it important to the test that these be in the scope of the timer callback? If not, then move them out, if so, then maybe this isn't an issue with the middleware, but how you're using the API.

File: rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp
Comment (wjwwood) [inline rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp:331]: This time is also likely to be flaky in the long term. There's really no value in not setting it to a higher value.

You could also use a future and `spin_until_future_complete`.

File: rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp
Comment (wjwwood) [inline rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp:333]: I don't think it's important for the test, but this method will add and remove the node from an executor each time this is called, so it would probably be better to create an executor, add the node, and either spin in a thread or use spin until future complete with a timeout, rather than using one of the `rclcpp::spin*` functions.

**Review by wjwwood — CHANGES_REQUESTED** [review]
I reexamined the test to check what @asorbini was asking about, and I think I found some flaws, but the general idea (i.e. using more than one executor per node via `add_callback_group()` and having a callback group wake an executor when something is added to it in order to consider new items) is a supported use case based on the API.

It might be that this is a bug in rclcpp and not the middlewares however. I need to look more deeply at the root cause of the issue. But I don't have time at the moment.

File: rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp
Comment (iuhilnehc-ynos) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp:324]: > Is it important to the test that these be in the scope of the timer callback?

I want to use the simplest source code to emulate the reported example of this issue.
timer callback ---> service callback
create a subscription on the service callback
there is a remote publisher to publish a message to the subscription,
that's all to reproduce this issue.

File: rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp
Comment (iuhilnehc-ynos) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp:333]: This kind of test code already existed in the `rclcpp` test, is it a problem here?

File: rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp
Comment (iuhilnehc-ynos) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp:324]: `100ms` might be caused this test case to be flaky, I'll update it.

Comment (wjwwood) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp]: > I want to use the simplest source code to emulate the reported example of this issue.

Yes, but do you understand that not keeping these in scope might cause issues? I don't think (I could be wrong) that is what the original problem was surrounding.

Also, sleeping in the timer callback will block anything else (either something you set up or functions of the node that are implementation details) from happening via callback (for a single threaded executor) so you should never do that if you can avoid it.

Comment (iuhilnehc-ynos) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp]: Please refer to https://github.com/ros2/rclcpp/issues/1611#issue-846709308 reported by user,

```
  void srv_callback(const std_srvs::srv::Trigger::Request::SharedPtr, std_srvs::srv::Trigger::Response::SharedPtr res){
    RCLCPP_INFO_STREAM(this->get_logger(), "Received request");
    new_msg_ = false;

    // Create subscription
    auto sub_options_ = rclcpp::SubscriptionOptions();
    sub_options_.callback_group = callback_group_;
    auto subscription = this->create_subscription<std_msgs::msg::Bool>(
          "topic_bool",
          rclcpp::QoS(1),
          std::bind(
            &CallbackGroupSrvServer::sub_callback,
            this,
            std::placeholders::_1),
          sub_options_
          );

    // Wait for new_msg
    while (rclcpp::ok()
           && new_msg_ == false) {
      rclcpp::sleep_for(std::chrono::seconds(1));
      RCLCPP_INFO_STREAM(this->get_logger(), "Sleep...");
    }

    res->message = "";
    res->success = true;
    return;
  }
```

`srv_callback` running on the main thread, and the `CallbackGroupSrvServer::sub_callback` running on the `callback_group_executor_thread_` thread, so it seems it's a normal case. (publishing message can run on the main thread), so I think it's not a problem.

Comment (iuhilnehc-ynos) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp]: And the user said `which we use a lot.`

Comment (iuhilnehc-ynos) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp]: I think that we'd better use the same case (but using three applications is unnecessary) to reproduce the issue by adding a regression test,

Comment (iuhilnehc-ynos) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp]: I don't want to argue about whether it is a good design or not. I mean It's just a use case somebody might use.

Comment (wjwwood) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp]: > I don't want to argue about whether it is a good design or not. I mean It's just a use case somebody might use.

Yes, but it is an anti-pattern (two of them):

- not keeping the pub/sub in scope and assuming some short period of time is long enough for the message to arrive
- blocking in a service/timer (or anything) callback

We could warn or prevent these if it were possible, but we cannot prevent the user from doing so or know when to warn and when not to.

This is a test, not some random user code, so we should avoid these issues. And doing so we should figure out what about the user reported code is the source of the problem. Is it the blocking in the callback? Is it the pub/sub going out of scope? Is it a bug in rclcpp? It's hard to tell without eliminating the obviously bad things to do.

File: rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp
Comment (wjwwood) [reply rclcpp/test/rclcpp/test_add_callback_groups_to_executor.cpp]: No, but it's not ideal either. The tests are pretty scattered in terms of age and quality, so they're not necessarily the best example for best practices.

Like I said, it's probably not affecting the test, but using a future and `spin_until_future_complete` is better. Also, since this test is to do with how nodes, callback groups, and executors interact, it's probably better to not add/remove the node to the executor in a loop...

**Comment by iuhilnehc-ynos** [comment]
This kind of test source code seems not good (adding it into rclcpp might make other users wrongly use it in the future), close it.

**Comment by fujitatomoya** [comment]
i do not know why we need to close this. after `rclcpp` is fixed, we can add this kind of test?

**Comment by wjwwood** [comment]
> i do not know why we need to close this. after rclcpp is fixed, we can add this kind of test?

I don't think we needed to close it, but I do think the things I pointed out needed to be addressed.

I do not think, however, that we should have tests that copy user submitted code by including anti-patterns or otherwise not-recommended practices. Instead, the test should (ideally) test one potential issue at a time. It can be distilled from a user reported issue, but again it should be cleaned up and written as properly as possible.

**Comment by fujitatomoya** [comment]
yeah okay that makes sense to me.

**Comment by iuhilnehc-ynos** [comment]
reopen it after not using some anti-pattern source code.

**Comment by iuhilnehc-ynos** [comment]
@wjwwood 

I have updated `rclcpp` source code by following https://github.com/ros2/rclcpp/issues/1611#issuecomment-843726301 if I understood correctly. 
Could you help to review it again?

File: rclcpp/src/rclcpp/executor.cpp
Comment (fujitatomoya) [inline rclcpp/src/rclcpp/executor.cpp:227]: just to confirm, this does not need to check if `weak_group_ptr` is already in the map, since it already checks with `get_associated_with_executor_atomic` above?

**Review by fujitatomoya — COMMENTED** [review]
overall, i think it looks good.

File: rclcpp/src/rclcpp/executor.cpp
Comment (iuhilnehc-ynos) [reply rclcpp/src/rclcpp/executor.cpp:227]: Yes

**Comment by fujitatomoya** [comment]
@wjwwood friendly ping.

**Comment by SteveMacenski** [comment]
Noting that this is causing a bug in the Humble migration and this fixes it for an issue in Nav2 linked above @clalancette

**Comment by clalancette** [comment]
It looks like this needs a rebase, and another review for the latest changes.  Once that is done, we can hopefully merge it in for Rolling.

Backporting it to Humble is going to be tricky, as it changes ABI.  We'll have to consider another way to do that.

**Comment by iuhilnehc-ynos** [comment]
@wjwwood friendly ping.

**Comment by SteveMacenski** [comment]
Also a friendly ping :-) @clalancette

**Comment by wjwwood** [comment]
Sorry for dropping this one.

So, what I was going to say last time I reviewed this, but then got distracted and didn't follow up (again sorry), is that this breaks API.

I think we should perhaps do a few things differently here:

- deprecate the existing API signature, and when it is used use the https://docs.ros2.org/galactic/api/rclcpp/namespacerclcpp_1_1contexts.html#a18a2108b667f7e582db2c9ef216870a6 function
  - This lets us tick-tock, as well as back port with minimal changes
- add new constructor with context argument (this pr already does this)
- consider adding a method to the node class (or a free function takes the node as an argument) which creates the callback group for you, so that users can create a callback group with a node rather than a context, in which case it uses the context from the node
  -  this is nice because then the user a) doesn't have to know what a context is, and b) the user isn't responsible for ensuring the same context is used for the callback group and the node (which is required)

Also that last point brings up the issue that we assume that the context used with the callback group and the entities later associated with the callback groups all use the context. We should probably assert that is the case somewhere.

**Review by wjwwood — CHANGES_REQUESTED** [review]
This lgtm, except for the breaking API without deprecation, which I mentioned in my most recent comment. I think if we address that, then we can move forward with this.

File: rclcpp/src/rclcpp/callback_group.cpp
Comment (wjwwood) [inline rclcpp/src/rclcpp/callback_group.cpp:24]: Specifically this change is breaking API.

**Comment by wjwwood** [comment]
Yet another option is that we could defer the creation of the guard condition for each callback group until they are added to an executor. At that time we could pass the context from the executor to the callback group, or even create the guard condition in the executor and give the guard condition to the callback group (dependency injection).

This approach prevents the callback group and executor's context from being different (which is not allowed) and it would also be possible to do without changing API, and maybe be easier to do without breaking ABI for the back ports.

**Comment by iuhilnehc-ynos** [comment]
> Yet another option is that we could defer the creation of the guard condition for each callback group until they are added to an executor. 

> It needs some more thought, but I actually like this approach better.

Thanks for the guide. I'll dig into it to see if it works.

**Comment by iuhilnehc-ynos** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17029/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11589/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17448/)

**Comment by iuhilnehc-ynos** [comment]
fix cpplint issue and rebased as system_tests needs the latest cpplint, re-CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17029/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11590/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17449/)

**Comment by iuhilnehc-ynos** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17031/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11592/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17451/)

File: rclcpp/src/rclcpp/callback_group.cpp
Comment (iuhilnehc-ynos) [reply rclcpp/src/rclcpp/callback_group.cpp]: removed it because of deferring to create the guard condition in the callback gruop

**Comment by iuhilnehc-ynos** [comment]
rerun CI for windows:

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17457/)

**Comment by iuhilnehc-ynos** [comment]
cc: @wjwwood

File: rclcpp/include/rclcpp/callback_group.hpp
Comment (wjwwood) [inline rclcpp/include/rclcpp/callback_group.hpp:181]: nitpick:

```suggestion
  create_notify_guard_condition(const rclcpp::Context::SharedPtr context_ptr);
```

File: rclcpp/src/rclcpp/callback_group.cpp
Comment (wjwwood) [inline rclcpp/src/rclcpp/callback_group.cpp:103]: It is technically ok as-is, but I'm curious why you used `make_unique` instead of `make_shared`?

Comment (wjwwood) [inline rclcpp/src/rclcpp/callback_group.cpp:103]: I don't love that this implicitly resets the notify guard condition if it was already set. It sounds like a subtle bug waiting to happen.

I kind of think that we should either have an interface like this:

- `create_notify_guard_condition()` -> throws if already has one
- `destroy_notify_guard_condition()` -> throws if not already set
- `get_notify_guard_condition()` -> could return nullptr

Or something like this:

- `exchange_notify_guard_condition()` -> returns previous pointer value, could be nullptr
  - the guard condition would be created by the executor and passed in here
- `get_notify_guard_condition()` -> can return nullptr
  - this could store/return a `weak_ptr`

Both of these are more explicit about changing controlling the lifecycle of the shared pointer.

File: rclcpp/include/rclcpp/callback_group.hpp
Comment (wjwwood) [inline rclcpp/include/rclcpp/callback_group.hpp:186]: This is weird to me because the guard condition is stored as a shared_ptr, so why return a weak_ptr?

**Review by wjwwood — CHANGES_REQUESTED** [review]

**Comment by wjwwood** [comment]
Thanks for iterating!

File: rclcpp/include/rclcpp/callback_group.hpp
Comment (iuhilnehc-ynos) [reply rclcpp/include/rclcpp/callback_group.hpp:181]: I think you asked me to remove the `const` as the constructor of `rclcpp::GuardCondition` uses a non-const context.

File: rclcpp/src/rclcpp/callback_group.cpp
Comment (iuhilnehc-ynos) [reply rclcpp/src/rclcpp/callback_group.cpp:103]: Sorry. I used the `unique_ptr` for the member at the beginning. For some reason, I changed to `shared_ptr` for it but forgot to update this code.

Comment (iuhilnehc-ynos) [reply rclcpp/src/rclcpp/callback_group.cpp:103]: The function is not intended to create the guard condition multiple times in the callback group. I'll update from `create_notify_guard_condition` to `get_notify_guard_condition`, and also add a check statement.

File: rclcpp/include/rclcpp/callback_group.hpp
Comment (iuhilnehc-ynos) [reply rclcpp/include/rclcpp/callback_group.hpp:186]: Because this get_notify_guard_condition could return nullptr as the creation function is a deferred function. (I don't know what kind of this case could be.)

- use shared_ptr

```cpp
auto gc = get_notify_guard_condition();
if (gc)
   gc->trigger();
```

- use weak_ptr

```cpp
if (auto gc = get_notify_guard_condition().lock() )
   gc->trigger();
```

so I decided to use the weak_ptr for this function.

File: rclcpp/src/rclcpp/callback_group.cpp
Comment (iuhilnehc-ynos) [reply rclcpp/src/rclcpp/callback_group.cpp]: Sorry, I didn't think over your suggestions before.
I think the first one you suggested is good. Thank you for the detailed guidance.

File: rclcpp/src/rclcpp/executor.cpp
Comment (iuhilnehc-ynos) [inline rclcpp/src/rclcpp/executor.cpp:695]: @wjwwood 

Is it worth introducing this code not to break the CallbackGroup API and make it defer to create the guard condition in the callback group?
I was worried about adding extra code in this kind of the main loop. I appreciate it if you have any suggestions.

**Comment by iuhilnehc-ynos** [comment]
retry CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17090/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11641/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17509/)

**Comment by iuhilnehc-ynos** [comment]
CI again:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17092/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11643/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17511/)

**Comment by SteveMacenski** [comment]
This all passed now!

File: rclcpp/src/rclcpp/executor.cpp
Comment (fujitatomoya) [reply rclcpp/src/rclcpp/executor.cpp:695]: @wjwwood friendly ping.

**Comment by fujitatomoya** [comment]
@wjwwood could you take an another look on your previous comments? i believe this has some good impact for ROS 2 system and user application.

**Comment by alsora** [comment]
We have been cherry-picking this PR in our Humble branch and it fixed the problem with callback groups

**Review by alsora — APPROVED** [review]

**Review by clalancette — CHANGES_REQUESTED** [review]
I rebased this PR and made a few minor changes (just to the include files).

Overall, I think the idea behind this PR is good, but I will point out that it is *not* thread-safe.  In particular, the lock around `get_notify_guard_condition` is useless; it is only held while manipulating the shared_ptr, but is then dropped when we return the object.  Anything we do with that object afterwards is not thread-safe.

What I think we should do instead is to make sure we do any operations on the callback-group guard condition with the lock held.  That would require us to add some additional APIs (like a new `trigger` method), modify some of the existing ones (change `destroy_notify_guard_condition` to only destroy the guard condition if it was created and not error out), and otherwise clean some things up.  With those things done, we should be able to make this thread-safe and move forward with this.

**Comment by iuhilnehc-ynos** [comment]
One callback group can only have one [notify_guard_condition_](https://github.com/ros2/rclcpp/blob/8dd0ec85a70412f707439dda4843ad0f91d7ddd0/rclcpp/include/rclcpp/callback_group.hpp#L235), the [notify_guard_condition_mutex_](https://github.com/ros2/rclcpp/blob/8dd0ec85a70412f707439dda4843ad0f91d7ddd0/rclcpp/include/rclcpp/callback_group.hpp#L236) is jsut for these [three operations](https://github.com/ros2/rclcpp/blob/8dd0ec85a70412f707439dda4843ad0f91d7ddd0/rclcpp/src/rclcpp/callback_group.cpp#L110-L139), and the callback group can be [used in only one executor](https://github.com/ros2/rclcpp/blob/8dd0ec85a70412f707439dda4843ad0f91d7ddd0/rclcpp/src/rclcpp/executor.cpp#L208-L212).

so it seems unnecessary to add extra mutex to make the callback-group guard condition thread-safe.

**Comment by iuhilnehc-ynos** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17533/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12068/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18088/)

**Comment by iuhilnehc-ynos** [comment]
retry CI:

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18098/)

**Comment by fujitatomoya** [comment]
@iuhilnehc-ynos thanks for the effort on this, i will take a look at this today. I think this is very helpful for user application.

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Comment by fujitatomoya** [comment]
dependent and above CI all green, i will go ahead to merge this.

**Comment by fujitatomoya** [comment]
@iuhilnehc-ynos i think we can backport this to humble, right? it would be really nice to backport since this affects user application behavior.

**Comment by fujitatomoya** [comment]
@Mergify backport this to humble

Just create backport PR so that we will not forget, let's give it a couple of weeks to make sure there is no degression, then we can merge the backport PR to humble.

**Pull request merged**
